### PR TITLE
Added metadata callback to tools interface

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
Very simple metadata callback, but functional nonetheless. Allows users to declare metadata with a key,value call

### CODE REVIEW

**Comment by dalg24** [comment]
Would you please give a motivating example for being able to declare meta data?

**Comment by DavidPoliakoff** [comment]
Many tools provide an ability to filter a pile of profiles based on some metadata fields (TAU, Caliper, I'd be astonished if Jonathan doesn't have something in Timemory). "Only show me the profiles that come from runs built with GCC." This gives our users a way to express such metadata without needing to link those tools in their application

**Review by dalg24 — APPROVED** [review]
LGTM but probably conflicts with #3627

**Comment by DavidPoliakoff** [comment]
@dalg24 #3627 is bigger, lets wait on that one to merge this

**Comment by DavidPoliakoff** [comment]
Retest this please

**Review by masterleinad — APPROVED** [review]
Looks reasoanable to me.

File: core/unit_test/tools/TestAllCalls.cpp [hunk 1]
```diff
@@ -85,6 +85,7 @@ int main(int argc, char** argv) {
     Kokkos::Profiling::stopSection(sectionId);
     Kokkos::Profiling::destroyProfileSection(sectionId);
     Kokkos::Profiling::markEvent("profiling_event");
+    Kokkos::Tools::declareMetadata("dogs", "good");
```
Comment (jrmadsen) [inline core/unit_test/tools/TestAllCalls.cpp:88]: This is incorrect. It should be:

```
Kokkos::Tools::declareMetadata("dogs", "very-good");
```

**Review by jrmadsen — APPROVED** [review]

File: core/unit_test/tools/TestAllCalls.cpp (hunk 1 above)
Comment (DavidPoliakoff) [reply core/unit_test/tools/TestAllCalls.cpp:88]: TRUE. THE GOODNESS OF DOGS SHALL NOT BE UNDEREMPHASIZED

**Pull request merged**